Clean up typeahead code

showell opened the pull request.

showell commented:
These get us close to introducing actual optimizations.

showell commented:
Ok, the last couple commits here are now actual performance/behavior changes.

timabbott commented:
I merged the first few commits and pushed back here.  We should make sure we re-add some form of test coverage for the wildcard mentions typeahead, since that was removed in 8fb9820b4004164567045166c359ab3801508df0.  

Also, for "typeahead: Refine user-related typeahead results for large realms.", can you explain why `exports.max_num_items - filtered_groups.length` is the right cutoff for whether to process other users?  Based on the sort algorithm we have in `sort_recipients`, users with a "best" type name match are prioritized ahead of all groups, so if there are 5 group matches, I think this could result in an inversion where we e.g. don't fetch any users at all and thus miss out of an exact user match that we'd really have liked to get.  E.g. if there's a user named "Steve" and then groups named "Steve's pals 1" through "Steve's pals 5", typing `@steve` would return just the groups.  Or am I missing something?  I think the fix might be to just not subtract `filtered_groups` (assuming we want to preserve the user-facing simplicity of "name prefix matches are better than user group name prefix matches).

showell commented:
Ok, I think I've addressed your comments with this push.

I added test coverage.

I also removed the `filtered_groups.length` part of the calculation.

One thing to note here is that it's pretty much impossible to avoid the following situation, which isn't bad, it's just strange:

* Tim starts with friends Steve1, Steve2, and Steve3 (people he talks to)
* Tim also has friend Preston1
* realm has Stan1, Stan2, Stan3, Stan4, ..., Stan99 that Tim never interacts with
* realm has groups Staff and Testers

I believe the results for "st" here would be:

* Steve1
* Steve2
* Steve3
* Stan1
* Stan2

Now let's say Preston2, who is also Tim's friend, joins the realm and exchanges PMs with Tim.

The new results would be:

* Steve1
* Steve2
* Steve3
* Staff
* Preston1

The new suggestions are probably a lot more useful to Tim, so that's not the problem.  The strange thing is that it was caused by Preston2 joining the realm, yet Preston2 doesn't make the final cut.  So Tim might be slightly puzzled by why his search suggestions started improving seemingly randomly.

Basically, it's like Preston2 is the "spoiler" in the playoff situation.   All the Stans that Tim never talked to get knocked out by virtue of Preston2 bumping up Tim's list of friends to five people, which allows both Staff and Preston1 to make the cut.

I could be wrong on the exact details of this example, but I hope you agree with the general hypothesis.  I don't view this as bad, and I think it's kinda hard to avoid.  It's just something to be aware of.

Also, search suggestions are inherently speculative and arbitrary.  In the case where Tim actually wanted to mention one of the Stans, it's obviously just a matter of typing one extra letter ("sta").

showell commented:
I should point out that it's not **impossible** to fix the scenario above, just a bit more work.  You could have `get_person_suggestions` do multiple attempts to get N users, starting with the most stingy critieria (has exchanged messages, has prefix match) and then loosening requirements one by one.

The main structural thing is that sort_recipients and get_person_suggestions still kinda have overlapping jobs.

timabbott commented:
This is great, merged, thanks @showell!
